### PR TITLE
tests: add stack usage metrics

### PR DESCRIPTION
**Pull request by kaspar030**
### Contribution description

This PR adds a module that, when enabled, will make threads print their stack usage *on exit*, in a format that the CI's metrics collection understands (and is somewhat human readable).

Example:

<details>

```
main(): This is RIOT! (Version: 2022.04-devel-501-g310e0-add_stack_usage_metrics)
If you can read this:                                                                                                                              
4294967295                                                               
-2147483648
FA
-2147483648
12345678
123456789ABCDEF0
18446744073709551615
-9223372036854775808
1.23450
Test successful.
{ "stack usage": { "main": 376 }}
```
</details>

This has *some* impact, the printing itself uses printf. Also, the actual stack data fields are kept per-thread, so 8 bytes more per thread. That means unfortunately some tests for some small-ram boards fall over the cliff.

### Testing procedure

CI should be fine.

### CODE REVIEW

**Comment by kaspar030** [comment]
this might still have some DEVELHELP related issues.

File: tests/Makefile.tests_common [hunk 1]
```diff
@@ -8,6 +8,10 @@ ifneq (,$(wildcard $(CURDIR)/tests*/.))
   endif
 
   DEFAULT_MODULE += test_utils_print_stack_usage
+  # add stack metric printing configuration when testing Kconfig
+  ifeq (1,$(TEST_KCONFIG))
+    KCONFIG_ADD_CONFIG += $(RIOTBASE)/tests/test_print_stack_usage.config
```
Comment (leandrolanzieri) [inline tests/Makefile.tests_common:13]: I think we could just have a single `test_defaults.config` file that enables all the test modules by default when compiling tests.

Comment (kaspar030) [reply tests/Makefile.tests_common:13]: will do!

**Comment by kaspar030** [comment]
Mmmh, the metric collection fails when an application runs multiple threads with the same name (or, restarts one of them). :(

I'm trying to come up with a solution that doesn't require the metrics collection [script](https://github.com/RIOT-OS/murdock-scripts/blob/master/post-build.d/01metrics.py) to deal with metric specific cases, like, `if metric=="stack usage"`.

Would creating a list make sense, if there's multiple entries? that would look like this:

```
   691         "tests/malloc_thread_safety": {                                                                                                     
   692             "native:gnu": {                                                                                                                 
   693                 "stack usage": {                                                                                                            
   694                     "idle": {                                                                                                               
   695                         "size": 8192,                                                                                                       
   696                         "used": 600                                                                                                         
   697                     },                                                                                                                      
   698                     "main": {                                                                                                               
   699                         "size": 12288,                                                                                                      
   700                         "used": 2508                                                                                                        
   701                     },                                                                                                                      
   702                     "t1": {                                                                                                                 
   703                         "size": 4096,                                                                                                       
   704                         "used": [                                                                                                           
   705                             860,                                                                                                            
   706                             472                                                                                                             
   707                         ]                                                                                                                   
   708                     },                                                                                                                      
   709                     "t2": {                                                                                                                 
   710                         "size": 4096,                                                                                                       
   711                         "used": [                                                                                                           
   712                             860,                                                                                                            
   713                             472                                                                                                             
   714                         ]                                                                                                                   
   715                     }                                                                                                                       
   716                 }                                                                                                                           
   717             }                                                                                                                               
   718         },
```

(note the multiple "used" entries for t1 and t2)

**Comment by kaspar030** [comment]
OK, I went with the threads being a list:

```
    1 {                                                                                                                                            
    2     "metrics": {                                                                                                                             
    3         "tests/malloc_thread_safety": {                                                                                                      
    4             "native:gnu": {                                                                                                                  
    5                 "threads": [                                                                                                                 
    6                     {                                                                                                                        
    7                         "name": "t1",                                                                                                        
    8                         "stack_size": 4096,                                                                                                  
    9                         "stack_used": 860                                                                                                    
   10                     },                                                                                                                       
   11                     {                                                                                                                        
   12                         "name": "t2",                                                                                                        
   13                         "stack_size": 4096,                                                                                                  
   14                         "stack_used": 860                                                                                                    
   15                     },                                                                                                                       
   16                     {                                                                                                                        
   17                         "name": "t1",                                                                                                        
   18                         "stack_size": 4096,                                                                                                  
   19                         "stack_used": 472                                                                                                    
   20                     },                                                                                                                       
   21                     {                                                                                                                        
   22                         "name": "t2",                                                                                                        
   23                         "stack_size": 4096,                                                                                                  
   24                         "stack_used": 472                                                                                                    
   25                     },                                                                                                                       
```

this needs a fix to the metrics script so lists get concatenated.

**Comment by MrKevinWeiss** [comment]
results after the murdock scripts have been updated

<details>

```json
{
    "metrics": {
        "tests/malloc_thread_safety": {
            "native:gnu": {
                "threads": [
                    {
                        "name": "t1",
                        "stack_size": 4096,
                        "stack_used": 860
                    },
                    {
                        "name": "t2",
                        "stack_size": 4096,
                        "stack_used": 860
                    },
                    {
                        "name": "t1",
                        "stack_size": 4096,
                        "stack_used": 472
                    },
                    {
                        "name": "t2",
                        "stack_size": 4096,
                        "stack_used": 472
                    },
                    {
                        "name": "idle",
                        "stack_size": 8192,
                        "stack_used": 600
                    },
                    {
                        "name": "main",
                        "stack_size": 12288,
                        "stack_used": 2508
                    }
                ]
            }
        },
        "tests/thread_cooperation": {
            "native:gnu": {
                "threads": [
                    {
                        "name": "thread",
                        "stack_size": 8192,
                        "stack_used": 2508
                    },
                    {
                        "name": "thread",
                        "stack_size": 8192,
                        "stack_used": 2508
                    },
                    {
                        "name": "thread",
                        "stack_size": 8192,
                        "stack_used": 2508
                    },
                    {
                        "name": "thread",
                        "stack_size": 8192,
                        "stack_used": 2508
                    },
                    {
                        "name": "thread",
                        "stack_size": 8192,
                        "stack_used": 2508
                    },
                    {
                        "name": "thread",
                        "stack_size": 8192,
                        "stack_used": 2508
                    },
                    {
                        "name": "thread",
                        "stack_size": 8192,
                        "stack_used": 2508
                    },
                    {
                        "name": "thread",
                        "stack_size": 8192,
                        "stack_used": 2412
                    },
                    {
                        "name": "thread",
                        "stack_size": 8192,
                        "stack_used": 2412
                    },
                    {
                        "name": "thread",
                        "stack_size": 8192,
                        "stack_used": 2412
                    },
                    {
                        "name": "thread",
                        "stack_size": 8192,
                        "stack_used": 2412
                    },
                    {
                        "name": "thread",
                        "stack_size": 8192,
                        "stack_used": 2412
                    },
                    {
                        "name": "idle",
                        "stack_size": 8192,
                        "stack_used": 436
                    },
                    {
                        "name": "main",
                        "stack_size": 12288,
                        "stack_used": 2612
                    }
                ]
            }
        }
    }
}
```
</details>

File: sys/test_utils/print_stack_usage/print_stack_usage.c [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+ * @author      Kaspar Schleiser <kaspar@schleiser.de>
+ */
+
+#include "thread.h"
+#include "log.h"
+
+void print_stack_usage_metric(const char *name, void *stack, unsigned max_size)
+{
+    unsigned free = thread_measure_stack_free(stack);
+    LOG_INFO("{ \"threads\": [{ \"name\": \"%s\", \"stack_size\": %u, \"stack_used\": %u }]}\n", name, max_size, max_size - free);
```
Comment (maribu) [inline sys/test_utils/print_stack_usage/print_stack_usage.c:26]:
```suggestion
    LOG_INFO("{ \"threads\": [{ \"name\": \"%s\", \"stack_size\": %u, \"stack_used\": %u }]}\n",
             name, max_size, max_size - free);
```

**Review by maribu — APPROVED** [review]
Code looks good to me and test output was provided. Please squash the style nitpick right in.

File: sys/test_utils/print_stack_usage/print_stack_usage.c (hunk 2 above)
Comment (kaspar030) [reply sys/test_utils/print_stack_usage/print_stack_usage.c]: done.

I wish github would just do "fixup" commits and provide "autosquash".

**Comment by kaspar030** [comment]
Thanks!

**Comment by fjmolinas** [comment]
`tests/pthread_flood` is failing on `samr21-xpro` after this PR, I'm guessing stack usage? Could you take a look @kaspar030 ?

```
47dd3b1889c813550c1029d03c06c404a29d42cb is the first bad commit
commit 47dd3b1889c813550c1029d03c06c404a29d42cb
Author: Kaspar Schleiser <kaspar@schleiser.de>
Date:   Thu Aug 6 13:27:13 2020 +0200

    tests: make test_utils_print_stack_usage a default module

 tests/Makefile.tests_common | 2 ++
 1 file changed, 2 insertions(+)
```

**Comment by kaspar030** [comment]
on it

**Comment by kaspar030** [comment]
Hm. The pthread threads' stack is too small for the printing of the stack sizes :/

**Comment by fjmolinas** [comment]
> Hm. The pthread threads' stack is too small for the printing of the stack sizes :/

DISABLE_MODULE then?

**Comment by kaspar030** [comment]
Using fmt if available, don't print anything if the stack is too small: https://github.com/RIOT-OS/RIOT/pull/17891

**Pull request merged**